### PR TITLE
Python 3 support

### PR DESCRIPTION
**Pull request by rlafuente**

### CODE REVIEW

**Comment by tian2992** [comment]
Does this lines https://github.com/shoebot/shoebot/blob/py3/.travis.yml#L7 in travis yml can be fixed?

**Comment by rlafuente** [comment]
> Does this lines https://github.com/shoebot/shoebot/blob/py3/.travis.yml#L7 in travis yml can be fixed?

Good eye Tian, thank you! I'm now trying to get it fixed through Stack Overflow answers :p

**Comment by rlafuente** [comment]
Okay we're still getting errors here, dropping it here so that Stu and me can look more closely at it later today:

1. commented out Vext, seems to be installing with py2 and complains on the Ruamel install that it's trying to install it system-wide
2. 2 unit tests are also failing, output follows.

```
======================================================================
ERROR: test_copy_len (tests.unittests.test_botapi_bezier.BezierTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/shoebot/shoebot/tests/unittests/test_botapi_bezier.py", line 13, in test_copy_len
    b1 = BezierPath(self.bot)
  File "/home/travis/build/shoebot/shoebot/shoebot/data/bezier.py", line 67, in __init__
    ColorMixin.__init__(self, fill=fill, stroke=stroke, strokewidth=strokewidth)
  File "/home/travis/build/shoebot/shoebot/shoebot/data/basecolor.py", line 257, in __init__
    self._set_strokewidth(strokewidth)
  File "/home/travis/build/shoebot/shoebot/shoebot/data/basecolor.py", line 293, in _set_strokewidth
    self._strokewidth = max(strokewidth, 0.0001)
TypeError: '>' not supported between instances of 'float' and 'NoneType'
======================================================================
ERROR: test_file_output (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_file_output
Traceback (most recent call last):
  File "/opt/python/3.7.1/lib/python3.7/unittest/loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
  File "/home/travis/build/shoebot/shoebot/tests/unittests/test_file_output.py", line 18, in <module>
    from util import TestSequence
ModuleNotFoundError: No module named 'util'
```

**Comment by rlafuente** [comment]
OK, I fixed the failing tests but there's still a problem with missing gi:

```
0.23s$ $PYTHON setup.py diagnose
running diagnose
Traceback (most recent call last):
  File "setup.py", line 197, in <module>
    test_suite="tests.unittests",
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/setuptools/__init__.py", line 145, in setup
    return distutils.core.setup(**attrs)
  File "/opt/python/3.7.1/lib/python3.7/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/opt/python/3.7.1/lib/python3.7/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/opt/python/3.7.1/lib/python3.7/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "setup.py", line 82, in run
    from shoebot.diagnose import diagnose
  File "/home/travis/build/shoebot/shoebot/shoebot/__init__.py", line 36, in <module>
    from shoebot.core.backend import cairo
  File "/home/travis/build/shoebot/shoebot/shoebot/core/__init__.py", line 39, in <module>
    from .cairo_canvas import CairoCanvas
  File "/home/travis/build/shoebot/shoebot/shoebot/core/cairo_canvas.py", line 35, in <module>
    from .backend import cairo
  File "/home/travis/build/shoebot/shoebot/shoebot/core/backend.py", line 142, in <module>
    driver = CairoGIBackend(get_driver_options())
  File "/home/travis/build/shoebot/shoebot/shoebot/core/backend.py", line 95, in __init__
    self.gi_lib, self.gi_module = self.setup_gi(gi_pref)
  File "/home/travis/build/shoebot/shoebot/shoebot/core/backend.py", line 104, in setup_gi
    name, gi_module = self.import_libs(['gi', 'pgi'], 'gi')
  File "/home/travis/build/shoebot/shoebot/shoebot/core/backend.py", line 63, in import_libs
    raise ImportError('No %s Implementation found, tried: %s' % (impl_name, ' '.join(module_names)))
ImportError: No gi Implementation found, tried: gi pgi
The command "$PYTHON setup.py diagnose" exited with 1.
```

There's also a vext error (i put it back on the Travis build file)

```
Building wheels for collected packages: vext.gi, vext
  Building wheel for vext.gi (setup.py) ... error
  Complete output from command /home/travis/virtualenv/python3.7.1/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-8faa31kt/vext.gi/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/pip-wheel-99nn9163 --python-tag cp37:
  running bdist_wheel
  running build
  installing to build/bdist.linux-x86_64/wheel
  running install
  vext.gi Install
  running install_data
  creating build
  creating build/bdist.linux-x86_64
  creating build/bdist.linux-x86_64/wheel
  creating build/bdist.linux-x86_64/wheel/vext.gi-0.7.0.data
  creating build/bdist.linux-x86_64/wheel/vext.gi-0.7.0.data/data
  copying gi.vext -> build/bdist.linux-x86_64/wheel/vext.gi-0.7.0.data/data/
  running install_egg_info
  running egg_info
  writing vext.gi.egg-info/PKG-INFO
  writing dependency_links to vext.gi.egg-info/dependency_links.txt
  writing requirements to vext.gi.egg-info/requires.txt
  writing top-level names to vext.gi.egg-info/top_level.txt
  reading manifest file 'vext.gi.egg-info/SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  writing manifest file 'vext.gi.egg-info/SOURCES.txt'
  Copying vext.gi.egg-info to build/bdist.linux-x86_64/wheel/vext.gi-0.7.0-py3.7.egg-info
  running install_scripts
  Install vext files:
  error: [Errno 2] No such file or directory: 'vext': 'vext'
  
  ----------------------------------------
  Failed building wheel for vext.gi
```

**Comment by stuaxo** [comment]
Nice :)

Some of these old files (looking at the SVG lib) have slightly weird formatting (the empty lines at the start of each function/method look odd to me... but maybe worth sorting another time.

Will try and have a look through in more detail + see if I can see anything.

File: lib/svg/__init__.py [hunk 1]
```diff
@@ -256,7 +256,13 @@ def parse_path(e):
         # Move the current point to the new coordinates.
         if command == "M":
             for i in range(int(len(points)/2)):
-                _ctx.moveto(points[i*2], points[i*2+1])
+                # More than one value is an implicit LINETO
+                # see https://www.w3.org/TR/SVG/paths.html#PathDataMovetoCommands
+                print(i)
```
Comment (stuaxo) [inline lib/svg/__init__.py]: Debug print statement left in here (and below)

File: lib/web/simplejson/__init__.py [hunk 2]
```diff
@@ -390,7 +390,7 @@ def main():
         raise SystemExit("%s [infile [outfile]]" % (sys.argv[0],))
     try:
         obj = load(infile)
-    except ValueError, e:
+    except ValueError(e):
```
Comment (stuaxo) [inline lib/web/simplejson/__init__.py]: Not sure I've seen this before, should it be `except ValueError as e` ?

File: lib/web/urbandictionary.py [hunk 3]
```diff
@@ -17,45 +17,45 @@ def __init__(self, word, url, description, example, author):
...
     pass
 
 class UrbanDictionary(list):
-    
+
     def __init__(self, q, cached=True):
-        
+
         url = "http://api.urbandictionary.com/soap?wsdl"
         key = "91cf66fb7f14bbf7fb59c7cf5e22155f"
```
Comment (stuaxo) [inline lib/web/urbandictionary.py]: OT - but, having keys in the code doesn't seem like the best idea - but API key management is almost a whole separate project in itself (something our IDE should maybe help with).

File: lib/web/urbandictionary.py [hunk 4]
```diff
@@ -17,45 +17,45 @@ def __init__(self, word, url, description, example, author):
...
+        # Live connect for uncached queries
         # or queries we do not have in cache.
         cache = Cache("urbandictionary", ".pickle")
         if not cached or not cache.exists(q):
             server = soap.SOAPProxy(url)
             try:
                 definitions = server.lookup(key, q)
-            except Exception, soap.faultType:
-                raise UrbanDictionaryError, "the API is no longer supported"
+            except Exception(soap.faultType):
```
Comment (stuaxo) [inline lib/web/urbandictionary.py]: I haven't seen this before, is it meant to be `except (Exception, soap.faultType):` ?

File: lib/web/url.py [hunk 5]
```diff
@@ -211,12 +211,12 @@ def open(url, wait=10):
                 o = urllib2.build_opener(p, urllib2.HTTPHandler)
                 urllib2.install_opener(o)
             connection = urllib2.urlopen(request)
```
Comment (stuaxo) [inline lib/web/url.py]: I think all these exception could be along the lines of `except urlib.XYZ as e`  ?

File: lib/graph/__init__.py [hunk 6]
```diff
@@ -52,12 +52,12 @@ def __init__(self, graph, id="", radius=8, style=style.DEFAULT, category="", lab
         self._betweenness = None
         self._eigenvalue = None
         
-        for k, v in properties.items():
+        for k, v in list(properties.items()):
```
Comment (stuaxo) [inline lib/graph/__init__.py]: Looks like this one might not need wrapping in a list, as it doesn't seem to try and modify `properties` ?

File: lib/graph/__init__.py [hunk 7]
```diff
@@ -158,7 +158,7 @@ def append(self, node, edge=None):
         list.append(self, node)
 
     def remove(self, node):
-        if self._edges.has_key(node.id): del self._edges[node.id]
+        if node.id in self._edges: del self._edges[node.id]
         list.remove(self, node)
```
Comment (stuaxo) [inline lib/graph/__init__.py]: Can all these deletion bits of code be fixed to use something atomic, like either -

```python
try:
    del self.edges[node.id]
except KeyError:
    pass
```

or 
`self.edges.pop(node.id, None)` ?

The whole of graph would probably improve readability with reformatting.

File: lib/graph/__init__.py [hunk 8]
```diff
@@ -357,7 +357,7 @@ def remove_edge(self, id1, id2):
     def node(self, id):
         """ Returns the node in the graph associated with the given id.
         """
-        if self.has_key(id):
+        if id in self:
             return self[id]
         return None
```
Comment (stuaxo) [inline lib/graph/__init__.py]: This could be simplified to `return self.get(id)`

File: lib/tuio/OSC.py [hunk 9]
```diff
@@ -263,14 +263,14 @@ def dispatch(self, message, source = None):
...
+            print('address %s not found, %s: %s' % (address, key, message))
             pprint.pprint(message)
-        except IndexError, e:
-            print '%s: %s' % (e, message)
+        except IndexError as e:
+            print('%s: %s' % (e, message))
             pass
-        except None, e:
-            print "Exception in", address, "callback :", e
+        except None as e:
```
Comment (stuaxo) [inline lib/tuio/OSC.py]: I had no idea you could raise None ?

**Comment by stuaxo** [comment]
Web library notes:  

It seems to vendor BeautifulSoup and simplejson, which is probably not the best way of doing things these days.

**Comment by stuaxo** [comment]
@rlafuente can we start using f-strings?   It should help with a lot of the places we do string manipulation, including in our nodebox-inherited libraries.

File: lib/colors/__init__.py [hunk 10]
```diff
@@ -952,10 +952,10 @@ def __init__(self, *args, **kwargs):
                         self.append(color(clr))
 
             # From a string (image/name/context).
-            if isinstance(arg, (str, unicode)):
+            if isinstance(arg, str):
                 if os.path.exists(arg):
                     n = 10
-                    if "n" in kwargs.keys(): n = kwargs["n"]
+                    if "n" in list(kwargs.keys()): n = kwargs["n"]
```
Comment (stuaxo) [inline lib/colors/__init__.py]: It doesn't look like this needs to be wrapped in a list (or the next two times either)

File: lib/colors/__init__.py [hunk 11]
```diff
@@ -2627,7 +2627,7 @@ def _weight_by_hue(self):
 
         # Calculate the normalized (0.0-1.0) weight for each hue,
         # and transform the dictionary to a list.
-        s = 1.0 * sum([w for r, w in grouped.values()])
+        s = 1.0 * sum([w for r, w in list(grouped.values())])
```
Comment (stuaxo) [inline lib/colors/__init__.py]: Looks like this doesn't need to be wrapped in a list.

File: lib/database/__init__.py [hunk 12]
```diff
@@ -271,7 +271,7 @@ def find(self, q, operator="=", fields="*", key=None):
         
         if key == None: key = self._key
         if fields != "*": fields = ", ".join(fields)
-        try: q = unicode(q)
+        try: q = str(q)
         except: pass
```
Comment (stuaxo) [inline lib/database/__init__.py]: I'm not sure about the wisdom of this exception handling, should we just do `q = str(q)` and be done with it ?

File: lib/database/__init__.py [hunk 13]
```diff
@@ -353,7 +353,7 @@ def edit(self, id, *args, **kw):
             fields = [k for k in kw]
             v = [kw[k] for k in kw]
         
-        sql  = "update "+self._name+" set "+"=?, ".join(fields)+"=? where "+self._key+"="+unicode(id)
+        sql  = "update "+self._name+" set "+"=?, ".join(fields)+"=? where "+self._key+"="+str(id)
         self._db._cur.execute(sql, v)
```
Comment (stuaxo) [inline lib/database/__init__.py]: Using fstrings here might make this all a bit more readable.

**Pull request merged**